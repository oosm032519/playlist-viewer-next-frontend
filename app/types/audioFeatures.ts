export type Mode = "MAJOR" | "MINOR";

export interface AudioFeatures {
    acousticness: number;
    danceability: number;
    energy: number;
    instrumentalness: number;
    liveness: number;
    speechiness: number;
    valence: number;
    tempo: number;
}

export interface TrackAudioFeatures extends AudioFeatures {
    durationMs: number;
    id: string;
    key: number;
    loudness: number;
    mode: Mode;
    timeSignature: number;
}
