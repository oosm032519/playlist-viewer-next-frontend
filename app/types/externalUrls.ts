// app/types/externalUrls.ts
export interface ExternalUrls {
    externalUrls: {
        spotify: string;
    };
}
