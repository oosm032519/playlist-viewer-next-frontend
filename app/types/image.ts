// app/types/image.ts

/**
 * 画像情報。
 */
export interface Image {
    height: number;
    url: string;
    width: number;
}
