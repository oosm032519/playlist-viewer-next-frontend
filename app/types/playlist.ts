export interface PlaylistImage {
    url: string;
}

export interface Playlist {
    tracks: any;
    id: string;
    name: string;
    description: string;
    images: PlaylistImage[];
}
