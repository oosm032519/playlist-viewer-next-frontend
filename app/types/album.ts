// app/types/album.ts
import {ExternalUrls} from '@/app/types/externalUrls'
import {Image} from '@/app/types/image'

export interface Album {
    name: string;
    images: Image[];
    externalUrls: ExternalUrls;
}
