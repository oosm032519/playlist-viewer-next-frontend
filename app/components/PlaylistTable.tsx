// app/components/PlaylistTable.tsx

import {Playlist} from "@/app/types/playlist";
import {Table, TableBody} from "@/app/components/ui/table";
import PlaylistTableHeader from "@/app/components/PlaylistTableHeader";
import PlaylistTableRow from "@/app/components/PlaylistTableRow";

/**
 * プレイリストテーブルコンポーネントのプロパティを定義するインターフェース
 */
interface PlaylistTableProps {
    /** プレイリストの配列 */
    playlists: Playlist[];
    /** プレイリストがクリックされたときのコールバック関数 */
    onPlaylistClick: (playlistId: string) => void;
    /** 現在のページ番号 */
    currentPage: number;
    /** 総プレイリスト数 */
    totalPlaylists: number;
}

/**
 * プレイリストのテーブルを表示するコンポーネント
 *
 * @param {PlaylistTableProps} props - プレイリストテーブルのプロパティ
 * @returns {JSX.Element} プレイリストテーブルのJSX要素
 *
 * @example
 * <PlaylistTable
 *   playlists={playlists}
 *   onPlaylistClick={handlePlaylistClick}
 *   currentPage={1}
 *   totalPlaylists={100}
 * />
 */
export default function PlaylistTable({
                                          playlists,
                                          onPlaylistClick,
                                          totalPlaylists,
                                      }: PlaylistTableProps): JSX.Element {
    return (
        <Table>
            {/* テーブルのヘッダー部分 */}
            <PlaylistTableHeader totalPlaylists={totalPlaylists}/> {/* totalPlaylists を渡す */}
            <TableBody>
                {/* プレイリストごとに行を生成 */}
                {playlists.map((playlist) => (
                    <PlaylistTableRow
                        key={playlist.id} // 各行に一意のキーを設定
                        playlist={playlist} // プレイリストのデータを渡す
                        onClick={() => onPlaylistClick(playlist.id)} // クリック時にコールバック関数を呼び出す
                    />
                ))}
            </TableBody>
        </Table>
    );
}
