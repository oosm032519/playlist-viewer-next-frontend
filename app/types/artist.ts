// app/types/artist.ts
import {ExternalUrls} from '@/app/types/externalUrls'

export interface Artist {
    name: string;
    externalUrls: ExternalUrls;
}
