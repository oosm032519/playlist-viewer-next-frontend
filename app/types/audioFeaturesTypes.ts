// app/types/audioFeaturesTypes.ts
import {Track} from '@/app/types/track';

export interface AudioFeaturesChartProps {
    track: Track | null;
}
