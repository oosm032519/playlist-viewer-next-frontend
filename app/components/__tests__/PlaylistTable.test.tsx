// PlaylistTable.test.tsx

import React from 'react';
import {render, screen, fireEvent} from '@testing-library/react';
import '@testing-library/jest-dom';
import PlaylistTable from '../PlaylistTable';
import {Playlist} from '@/app/types/playlist';

// モックデータの作成
const mockPlaylists: Playlist[] = [
    {
        id: '1',
        name: 'Playlist 1',
        images: [{url: 'https://example.com/image1.jpg'}],
        tracks: {total: 10},
        description: ''
    },
    {
        id: '2',
        name: 'Playlist 2',
        images: [{url: 'https://example.com/image2.jpg'}],
        tracks: {total: 20},
        description: ''
    },
];

describe('PlaylistTable', () => {
    // onPlaylistClickのモック関数を作成
    const mockOnPlaylistClick = jest.fn();
    
    // 各テストの前にモック関数をリセット
    beforeEach(() => {
        mockOnPlaylistClick.mockClear();
    });
    
    it('renders the table headers correctly', () => {
        render(<PlaylistTable playlists={mockPlaylists} onPlaylistClick={mockOnPlaylistClick}/>);
        
        expect(screen.getByText('Image')).toBeInTheDocument();
        expect(screen.getByText('Name')).toBeInTheDocument();
        expect(screen.getByText('Tracks')).toBeInTheDocument();
    });
    
    it('renders the correct number of playlists', () => {
        render(<PlaylistTable playlists={mockPlaylists} onPlaylistClick={mockOnPlaylistClick}/>);
        
        const playlistRows = screen.getAllByRole('row').slice(1); // ヘッダー行を除外
        expect(playlistRows).toHaveLength(mockPlaylists.length);
    });
    
    it('displays playlist information correctly', () => {
        render(<PlaylistTable playlists={mockPlaylists} onPlaylistClick={mockOnPlaylistClick}/>);
        
        mockPlaylists.forEach((playlist) => {
            expect(screen.getByText(playlist.name)).toBeInTheDocument();
            expect(screen.getByText(playlist.tracks.total.toString())).toBeInTheDocument();
            const image = screen.getByAltText(playlist.name) as HTMLImageElement;
            expect(image).toBeInTheDocument();
            expect(image.src).toBe(playlist.images[0].url);
        });
    });
    
    it('calls onPlaylistClick with correct playlist ID when a row is clicked', () => {
        render(<PlaylistTable playlists={mockPlaylists} onPlaylistClick={mockOnPlaylistClick}/>);
        
        const playlistRows = screen.getAllByRole('row').slice(1); // ヘッダー行を除外
        fireEvent.click(playlistRows[0]);
        expect(mockOnPlaylistClick).toHaveBeenCalledWith(mockPlaylists[0].id);
        
        fireEvent.click(playlistRows[1]);
        expect(mockOnPlaylistClick).toHaveBeenCalledWith(mockPlaylists[1].id);
    });
    
    it('renders empty table when no playlists are provided', () => {
        render(<PlaylistTable playlists={[]} onPlaylistClick={mockOnPlaylistClick}/>);
        
        const playlistRows = screen.getAllByRole('row').slice(1); // ヘッダー行を除外
        expect(playlistRows).toHaveLength(0);
    });
    
    it('applies correct CSS classes to table elements', () => {
        render(<PlaylistTable playlists={mockPlaylists} onPlaylistClick={mockOnPlaylistClick}/>);
        
        const images = screen.getAllByRole('img');
        images.forEach((img) => {
            expect(img).toHaveClass('w-12', 'h-12', 'object-cover', 'rounded-full');
        });
    });
});
