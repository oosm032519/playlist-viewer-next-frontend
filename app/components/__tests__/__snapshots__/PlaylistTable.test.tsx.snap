// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistTable matches snapshot 1`] = `
<div>
  <div
    class="relative w-full overflow-auto"
  >
    <table
      class="w-full caption-bottom text-sm"
    >
      <thead
        class="[&_tr]:border-b"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Image
          </th>
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Name
          </th>
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Tracks
          </th>
        </tr>
      </thead>
      <tbody
        class="[&_tr:last-child]:border-0"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            <img
              alt="Playlist 1"
              class="w-12 h-12 object-cover rounded-full"
              height="48"
              src="https://example.com/image1.jpg"
              width="48"
            />
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Playlist 1
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            10
          </td>
        </tr>
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            <img
              alt="Playlist 2"
              class="w-12 h-12 object-cover rounded-full"
              height="48"
              src="https://example.com/image2.jpg"
              width="48"
            />
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Playlist 2
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            20
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
